### DISCUSSION TITLE
columnar appender

### DISCUSSION THREAD

**Discussion by danz68** [duckdb/duckdb#3183] [Ideas]
Hello,

DuckDB uses a row based appender which makes sense for converting external row based storage data into internal format. But most of the data is retrieved in batches, therefore they can be sent to the duckdb engine in column batches as well. This would save a lot of API and internal calls.

duckdb_col_appender col_appender;
duckdb_col_appender_create(con, schema, table, 100000 /* max_batch_size */, & col_appender);
int val32[] = {1,2,3,4}
char **valvchar = {“first”, “second”, “third”, “forth”}
duckdb_col_append_int32(col_appender, val32, 4); 
duckdb_col_append_varchar(col_appender, valvchar,4);
duckdb_col_append_end( col_appender, 4);

**Comment by Mytherin** [comment]
I am working on something similar to this for the C API, but more generic. Specifically I made an [API to construct data chunks directly from C](https://github.com/Mytherin/duckdb/blob/julia/src/include/duckdb.h#L944), which can be done quickly in a columnar manner by getting the pointer directly. The data chunks can then be appended to the appender (see `duckdb_append_data_chunk`).

**Reply by danz68** [reply]
This is even better. Cant wait to see how it performs. Thanks! 

Talking about chunks. I was wondering if this can be also provided for results, as an alternate unidirectional extraction, considering that right now I can either pick one value at a time (duckdb_value - slow) or all at once (duckdb_column_data - can be huge). 

This mechanism can be tuned to provide even a true zero copy mechanism if the existing internal chunk is provided (if it exists, of course).

duckdb_create_chunked_buffer_for_result(result, int* col_idxs, int col_cnt /* zero for default buffer */, size_t *row_cnt /* if *row_cnt = zero, default or result row_cnt if smaller, returning the chunk size*/)
size_t duckdb_next_chunk(result, chunk) returning validated row_cnt of the chunk.

The rest of the API remains valid, as far as I can see.

**Reply by Mytherin** [reply]
Yes I intend to support this for result set consumption as well, although with a slightly different API.

**Reply by danz68** [reply]
I couldn't stop but notice the "duckdb_table_function" API. I assume it is a read only data mechanism, similar with table functions (C - used myself, R, Python API) in MonetDB. 

I've used in the past MonetDBLite with onlfly generated table functions to populate native tables and the only speed limitation was given by my own table implementation (I was reading from several different databases, different providers, etc). And yes, the trick was to hack the MonetDBLite CAPI code to accept not only on fly compilated functions, but also functions already compiled in DLLs, or even directly passed as pointers (I was using the inprocess library, afterall). This helped me to have not only onfly generated table functions, but I could also register new functions pointing to some inprocess functions.  

If I assume correctly,  the "duckdb_table_function" API goal is the other way around. You provide a mechanism of passing a pointer to a table function, and is up to the developer to extend whatever in top of it, including on fly compiled functions and such. This would be awesome. 

I was wondering if the "duckdb_table_function" API is going to evolve into full virtual tables API or it will remain on the raw data source level only. And, of course, I am wondering if  a "duckdb_scalar_function" API will follow as well. Because if it does, the CAPI mechanism will allow the development in any other language able to pass on-fly pointers (Delphi, Julia, Rust, YouNameIt) . 
However, because the C API cannot use C++ templates, just out of curiosity I made a quick math to see how many different functions with three parameter plus one return type (from 19 basic types in CAPI right now) one can make to be passed directly to duckdb engine. The result is ...  a discouraging 395,010 combinations!! Unless, of course, you have other ideas, I see no way one would be able to directly pass a function from C to duckdb engine and use it as it would be a native scalar function. 

(I don't dare to suggest anything, I just love this project and sometimes I love to have a chat about it.)
